File: server/utils/pipelineResolver.ts
import { prisma } from '../prismaClient';

const stageAliases = new Map<string, string>([
  ['Intake', 'Submitted'],
  ['Submitted', 'Submitted'],
  ['Classification', 'Classified'],
  ['Classified', 'Classified'],
  ['Qualification', 'Qualified'],
  ['Qualified', 'Qualified'],
  ['Scoring', 'Scored'],
  ['Scored', 'Scored'],
  ['Discovery', 'Discovery'],
  ['PDD Creation', 'PDD Creation'],
  ['PRD Creation', 'PDD Creation'],
  ['SDD Creation', 'SDD Creation'],
  ['Solution Design', 'SDD Creation'],
  ['Solution Designed', 'SDD Creation'],
  ['ROI', 'ROI Approved'],
  ['ROI Approved', 'ROI Approved'],
  ['Prioritization', 'Prioritized'],
  ['Prioritized', 'Prioritized'],
  ['Pod Allocation', 'Pod Allocated'],
  ['Pod Allocated', 'Pod Allocated'],
  ['Sprint Readiness', 'Sprint Ready'],
  ['Sprint Ready', 'Sprint Ready'],
]);

const normalizeStage = (stage: string) => stageAliases.get(stage) ?? stage;

export async function resolveNextStage(currentStage: string): Promise<string | null> {
  const stages = await prisma.stageConfig.findMany({
    orderBy: { order: 'asc' },
  });
  const normalizedCurrentStage = normalizeStage(currentStage);
  const currentIndex = stages.findIndex(s => normalizeStage(s.name) === normalizedCurrentStage);

  if (currentIndex === -1) {
    return stages.find(s => s.isEnabled)?.name ?? null;
  }

  return stages.slice(currentIndex + 1).find(s => s.isEnabled)?.name ?? null;
}

File: backend-dotnet/Services/PipelineResolver.cs
using EnterpriseAutomation.Api.Data;
using Microsoft.EntityFrameworkCore;

namespace EnterpriseAutomation.Api.Services;

public sealed class PipelineResolver(AppDbContext db)
{
    private static readonly Dictionary<string, string> StageAliases = new()
    {
        ["Intake"] = "Submitted",
        ["Submitted"] = "Submitted",
        ["Classification"] = "Classified",
        ["Classified"] = "Classified",
        ["Qualification"] = "Qualified",
        ["Qualified"] = "Qualified",
        ["Scoring"] = "Scored",
        ["Scored"] = "Scored",
        ["Discovery"] = "Discovery",
        ["PDD Creation"] = "PDD Creation",
        ["PRD Creation"] = "PDD Creation",
        ["A2B"] = "A2B Readiness Check",
        ["A2B Readiness Check"] = "A2B Readiness Check",
        ["SDD Creation"] = "SDD Creation",
        ["Solution Design"] = "SDD Creation",
        ["Solution Designed"] = "SDD Creation",
        ["ROI"] = "ROI Approved",
        ["ROI Approved"] = "ROI Approved",
        ["Prioritization"] = "Prioritized",
        ["Prioritized"] = "Prioritized",
        ["Pod Allocation"] = "Pod Allocated",
        ["Pod Allocated"] = "Pod Allocated",
        ["Sprint Readiness"] = "Sprint Ready",
        ["Sprint Ready"] = "Sprint Ready",
    };

    public static string Normalize(string stage) => StageAliases.GetValueOrDefault(stage, stage);

    public async Task<string?> ResolveNextStageAsync(string currentStage)
    {
        var stages = await db.StageConfigs.OrderBy(s => s.Order).ToListAsync();
        var normalized = Normalize(currentStage);
        var currentIndex = stages.FindIndex(s => Normalize(s.Name) == normalized);
        if (currentIndex == -1) return stages.FirstOrDefault(s => s.IsEnabled)?.Name;
        return stages.Skip(currentIndex + 1).FirstOrDefault(s => s.IsEnabled)?.Name;
    }
}
